### DISCUSSION TITLE
Optional plugins

### DISCUSSION THREAD

**Discussion by horseinthesky** [wbthomason/packer.nvim#229] [Q&A] [answered]
Hello.
I've just started the transition from VIM script to lua and decided to start with plugin manage. The first two questions come to mind:
1) As fas as I understood there are `optional` plugins and `start` plugins. But I could find the difference between them. Why should I install packer as opt plugin? Or in general, should I? The same is about start plugins.
2) I have some install options in VIm Plug:
```
Plug 'junegunn/fzf', { 'dir': '~/.fzf', 'do': './install --all' }
Plug 'junegunn/fzf.vim'
Plug 'kkoomen/vim-doge', { 'do': { -> doge#install() } }
```
How can I achieve the same in packer?
3) While doing the transition I want to my lua setup be loaded only for NVIM 0.5 so I added this to my init.vim
```
if has('nvim-0.5')
  :lua require'setup'
endif                
```

and then in my setup.lua I have
```
local fn = vim.fn
local execute = vim.api.nvim_command

-- Sensible defaults
require('settings')

-- Auto install packer.nvim if not exists
local install_path = fn.stdpath('data')..'/site/pack/packer/opt/packer.nvim'
if fn.empty(fn.glob(install_path)) > 0 then
  execute('!git clone https://github.com/wbthomason/packer.nvim '..install_path)
end
vim.cmd [[packadd packer.nvim]]
vim.cmd 'autocmd BufWritePost plugins.lua PackerCompile' -- Auto compile when there are changes in plugins.lua

-- Install plugins
require('plugins')

-- Setup Lua language server using submodule
require('lsp')
```

But whathever NVIM 0.4.3 or 0.5 I run it always loads packer (which in turn loads lsp etc.)
```
Invalid Neovim version for packer.nvim!
Error detected while processing /home/horseinthesky/.local/share/nvim/site/pack/packer/start/nvim-lspconfig/plugin/ls
pconfig.vim:
line   11:
E5105: Error while calling lua chunk: .../pack/packer/start/nvim-lspconfig/lua/lspconfig/util.lua:11: attempt to inde
x local 'lsp' (a nil value)
```

Is there a way to achieve what I want - packer only for NVIM 0.5?

Thank you.

**Comment by wbthomason** [comment, marked as answer]
> But I could find the difference between them. Why should I install packer as opt plugin? Or in general, should I? The same is about start plugins.

If you haven't read `:help packages`, I'd suggest starting there. The short version is that `start` plugins are eagerly loaded, whereas `opt` plugins must be loaded with `packadd`. A large part of `packer`'s job is handling the latter for you. In general, default to `start`, and use `opt` when you have some reason to lazy-load a plugin (e.g. it's only relevant for some filetypes, or imposes a high startup cost, etc.)

> How can I achieve the same in packer?

The equivalent is `packer`'s `run` key: https://github.com/wbthomason/packer.nvim#updateinstall-hooks

> Is there a way to achieve what I want - packer only for NVIM 0.5?

Hmm, not super easily. The problem is that `start` packages always get eagerly loaded, so you'd need to uninstall them between version switches. You could maybe do some trickery with setting `packpath` differently based on whether you're in 0.5 or not, but I haven't tried this myself and am not sure if it would work.

I hope this helps!

**Reply by horseinthesky** [reply]
Thank you.
I've opened https://github.com/wbthomason/packer.nvim/issues/237 and @nanotee and @akinsho helped me to find a more or less simple way to achieve what I wanted:
1) To use a custom path for `packer_compiled.vim`
2) To use a simple `cond`  to make plugins optional but still have lazy-loading for them.

There are two more related issues though I cannot figure out how to handle
https://github.com/wbthomason/packer.nvim/issues/244
https://github.com/wbthomason/packer.nvim/issues/245

Could you pls take a look at them?

**Reply by wbthomason** [reply]
Yep, I'm getting there. Sorry for the delay; I had to take a while fully away from `packer` for real work, so I'm working through the backlog of issues now.

**Reply by horseinthesky** [reply]
There is no hurry. I'm just investigating Lua's possibilities for my NVIM environment.
And want to say thank you help and such an interesting plugin manager =)
